src/metaxy/metadata_store/delta.py

```diff
@@ -0,0 +1,486 @@
+            deltalake.write_deltalake(
+                migration_events_uri,
+                empty_df,
+                mode="append",
+                storage_options=self.storage_options or None,
+            )
+
+    def _table_name_to_feature_key(self, table_name: str) -> FeatureKey:
+        """Convert table name back to feature key.
+
+        Args:
+            table_name: Table name (directory name) to parse
+
+        Returns:
+            FeatureKey constructed from table name parts
+        """
+        # Table names are created by joining parts with "__"
+        parts = table_name.split("__")
+        # Restore system table prefix (table_name converts hyphens to underscores)
+        if parts and parts[0] == "metaxy_system":
+            parts[0] = "metaxy-system"
+        return FeatureKey(parts)
+
+    def _feature_uri(self, feature_key: FeatureKey) -> str:
+        """Return the URI/path used by deltalake for this feature."""
+        if self.layout == "nested":
+            # Nested layout: store in directories like "part1/part2/part3"
+            table_path = "/".join(feature_key.parts)
+        else:
+            # Flat layout (default): store in directories like "part1__part2__part3"
+            table_path = feature_key.table_name
+        return os.path.join(self._root_uri, table_path)
+
+    def _feature_local_path(self, feature_key: FeatureKey) -> Path | None:
+        """Return filesystem path when operating on local roots."""
+        if self._local_root_path is None:
+            return None
+        if self.layout == "nested":
+            # Nested layout: use Path objects to create nested directories
+            return self._local_root_path / Path(*feature_key.parts)
+        else:
+            # Flat layout (default)
+            return self._local_root_path / feature_key.table_name
+
+    def _table_exists(self, table_uri: str) -> bool:
+        """Check whether the provided URI already contains a Delta table.
+
+        Works for both local and remote (object store) paths.
+        """
+        import deltalake  # pyright: ignore[reportMissingImports]
+
+        return deltalake.DeltaTable.is_deltatable(
+            table_uri,
+            storage_options=self.storage_options or None,
+        )
+
+    # ===== Storage operations =====
+
+    def write_metadata_to_store(
+        self,
+        feature_key: FeatureKey,
+        df: Frame,
+    ) -> None:
+        """Append metadata to the Delta table for a feature.
+
+        Args:
+            feature_key: Feature key to write to
+            df: DataFrame with metadata (already validated)
+
+        Raises:
+            TableNotFoundError: If table doesn't exist and auto_create_tables is False
+        """
+        import deltalake  # pyright: ignore[reportMissingImports]
+
+        from metaxy._utils import collect_to_polars
+
+        # Convert Narwhals DataFrame to Polars for delta-rs
+        df_polars = collect_to_polars(df)
```

Comment:
This will collect lazy frames, we need a slightly different thing: convert to Polars without collecting lazy frames. There is another helper in _utils.py that does that.